src/CMS/Search/SearchForm.php

```diff
@@ -37,4 +41,79 @@ public function classesToSearch($classes)
 		$this->classesToSearch = $legalClasses;
 	}
 
+    /**
+     * Return dataObjectSet of the results using current request to get info from form.
+     * Wraps around {@link searchEngine()}.
+     *
+     * @return SS_List
+     */
+    public function getResults()
+    {
+        // Get request data from request handler
+        $request = $this->getRequestHandler()->getRequest();
+
+        // set language (if present)
+        $locale = null;
+        $origLocale = null;
+        if (class_exists('Translatable')) {
+            $locale = $request->requestVar('searchlocale');
+            if (SiteTree::singleton()->hasExtension('Translatable') && $locale) {
+                if ($locale === "ALL") {
+                    Translatable::disable_locale_filter();
+                } else {
+                    $origLocale = Translatable::get_current_locale();
+
+                    Translatable::set_current_locale($locale);
+                }
+            }
+        }
+
+        $keywords = $request->requestVar('Search');
+
+        $andProcessor = function ($matches) {
+            return ' +' . $matches[2] . ' +' . $matches[4] . ' ';
+        };
+        $notProcessor = function ($matches) {
+            return ' -' . $matches[3];
+        };
+
+        $keywords = preg_replace_callback('/()("[^()"]+")( and )("[^"()]+")()/i', $andProcessor, $keywords);
+        $keywords = preg_replace_callback('/(^| )([^() ]+)( and )([^ ()]+)( |$)/i', $andProcessor, $keywords);
+        $keywords = preg_replace_callback('/(^| )(not )("[^"()]+")/i', $notProcessor, $keywords);
+        $keywords = preg_replace_callback('/(^| )(not )([^() ]+)( |$)/i', $notProcessor, $keywords);
+
+        $keywords = $this->addStarsToKeywords($keywords);
+
+        $pageLength = $this->getPageLength();
+        $start = $request->requestVar('start') ?: 0;
+
+        $booleanSearch =
+            strpos($keywords, '"') !== false ||
+            strpos($keywords, '+') !== false ||
+            strpos($keywords, '-') !== false ||
+            strpos($keywords, '*') !== false;
+        $results = DB::get_conn()->searchEngine($this->classesToSearch, $keywords, $start, $pageLength, "\"Relevance\" DESC", "", $booleanSearch);
+
+        // filter by permission
+        if ($results) {
+            foreach ($results as $result) {
+                if (!$result instanceof DataObject || !$result->canView()) {
```

Comment:
This is the particular fix, unfortunately we need the entire method.

Reply:
Hello @muskie9 

I am trying to understand why would the searchEngine function will return anything other than a data object.  If you look at this https://github.com/SilverStripers/elemental-seach/blob/master/src/ORM/Connect/MySQLDatabase.php

```
 foreach ($records as $record) {
            $object = DataList::create($record['ClassName'])->byID($record['ID']);
            if ($object && $object->canView()) {
                $objects[] = $object;
            }
        }

        $list = new PaginatedList(new ArrayList($objects));
        $list->setPageStart($start);
        $list->setPageLength($pageLength);
        $list->setTotalItems($totalCount);

        // The list has already been limited by the query above
        $list->setLimitItems(false);

        return $list;

```

If queries the database for the item and returns a data object and also check the particular objects canView function too. 

Am I missing something here ?

Reply:
I'll try to track down the edge case, but the error I was finding is in the `SearchForm`. The code you've shown _should_ result in all objects, however I did have at least one result that was `null` in the list for some reason. I'll see what I can find and report back.

Reply:
Ah, it may be that my local was pulling in a different vcs by way of private packagist. The latest master is in fact working as expected. I'll close this and the issue ticket.

Reply:
Cool thanks @muskie9